src/Compilers/Test/Core/Assert/ArtifactUploadUtil.cs

```diff
@@ -46,28 +52,57 @@ public void SetSucceeded()
 
         public void Dispose()
         {
-            if (!_success)
+            if (_success)
             {
-                var uploadDir = Environment.GetEnvironmentVariable("HELIX_WORKITEM_UPLOAD_ROOT");
-                if (string.IsNullOrEmpty(uploadDir))
+                return;
+            }
+
+            var uploadDir = Environment.GetEnvironmentVariable("HELIX_WORKITEM_UPLOAD_ROOT");
+            if (string.IsNullOrEmpty(uploadDir))
+            {
+                _testOutputHelper.WriteLine("Skipping artifact upload as not running in helix");
+                _testOutputHelper.WriteLine("Files");
+                foreach (var filePath in _filePaths)
                 {
-                    _testOutputHelper.WriteLine("Skipping artifact upload as not running in helix");
-                    _testOutputHelper.WriteLine("Artifacts");
-                    foreach (var filePath in _artifactFilePath)
-                    {
-                        _testOutputHelper.WriteLine(filePath);
-                    }
+                    _testOutputHelper.WriteLine(filePath);
+                }
+
+                _testOutputHelper.WriteLine("Directories");
+                foreach (var directory in _directoryPaths)
+                {
+                    _testOutputHelper.WriteLine(directory);
                 }
-                else
+            }
+            else
+            {
+                uploadDir = Path.Combine(uploadDir, _baseDirectoryName);
+                Directory.CreateDirectory(uploadDir);
+                _testOutputHelper.WriteLine($"Uploading artifacts by copying to {uploadDir}");
+                foreach (var filePath in _filePaths)
                 {
-                    uploadDir = Path.Combine(uploadDir, _baseDirectoryName);
-                    Directory.CreateDirectory(uploadDir);
-                    _testOutputHelper.WriteLine($"Uploading artifacts by copying to {uploadDir}");
-                    foreach (var filePath in _artifactFilePath)
+                    _testOutputHelper.WriteLine($"Copying file {filePath}");
```

Comment:
Does this need to deduplicate if this target file already exists?

Reply:
Rather than dedup it should error IMHO. If you're trying to debug a test failure by double copying files with the same name that's not great.